package org.gluu.oxtrust.exception;

import java.util.Iterator;

import jakarta.enterprise.context.NonexistentConversationException;
import jakarta.faces.FacesException;
import jakarta.faces.context.ExceptionHandler;
import jakarta.faces.context.ExceptionHandlerWrapper;
import jakarta.faces.context.ExternalContext;
import jakarta.faces.context.FacesContext;
import jakarta.faces.event.ExceptionQueuedEvent;
import jakarta.faces.event.ExceptionQueuedEventContext;

import org.apache.commons.lang.exception.ExceptionUtils;
import org.gluu.oxtrust.security.Identity;
import org.gluu.service.cdi.util.CdiUtil;
import org.gluu.service.security.SecurityEvaluationException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Created by eugeniuparvan on 5/23/17.
 */
public class GlobalExceptionHandler extends ExceptionHandlerWrapper {
    private Logger log = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    private ExceptionHandler wrapped;

	private boolean showJsfErrors = false;

    GlobalExceptionHandler(ExceptionHandler exception) {
		String enableJsfErrors = System.getProperties().getProperty("gluu.enable.jsf.errors");
		if ((enableJsfErrors != null) && Boolean.valueOf(enableJsfErrors)) {
			this.showJsfErrors  = true;
		}
        this.wrapped = exception;
    }

    @Override
    public ExceptionHandler getWrapped() {
        return this.wrapped;
    }

    public void handle() throws FacesException {
        final Iterator<ExceptionQueuedEvent> i = getUnhandledExceptionQueuedEvents().iterator();

        boolean renderResponse = true;
        while (i.hasNext()) {
            ExceptionQueuedEvent event = i.next();
            ExceptionQueuedEventContext context = (ExceptionQueuedEventContext) event.getSource();

            Throwable t = context.getException();
            final FacesContext fc = FacesContext.getCurrentInstance();
            final ExternalContext externalContext = fc.getExternalContext();
            try {
				if (isSecurityException(t)) {
					performRedirect(externalContext, "/login.htm");
				} else if (isConversationException(t)) {
					log.trace(t.getMessage(), t);
					performRedirect(externalContext, "/conversation_error.htm");
				} if (isViewExpiredException(t)) {
                    storeRequestURI();
                    performRedirect(externalContext, "/login.htm");
				} else {
					if (this.showJsfErrors) {
						renderResponse = false;
					} else {
						log.debug(t.getMessage(), t);
						performRedirect(externalContext, "/error.htm");				}
					}
				if (renderResponse) {
					fc.renderResponse();
				}
            } finally {
            	if (renderResponse) {
            		i.remove();
            	}
            }
        }
        getWrapped().handle();
    }

    protected void storeRequestURI() {
        ExternalContext extContext = FacesContext.getCurrentInstance().getExternalContext();
        String requestUri = ((jakarta.servlet.http.HttpServletRequest) extContext.getRequest()).getRequestURI();

        Identity identity = CdiUtil.bean(Identity.class);
        identity.setSavedRequestUri(requestUri);
    }

    private boolean isSecurityException(Throwable t) {
        return ExceptionUtils.getRootCause(t) instanceof SecurityEvaluationException;
    }

    private boolean isConversationException(Throwable t) {
        return ExceptionUtils.getRootCause(t) instanceof NonexistentConversationException;
    }

    private boolean isViewExpiredException(Throwable t) {
        return t instanceof jakarta.faces.application.ViewExpiredException;
    }

    private void performRedirect(ExternalContext externalContext, String viewId) {
        try {
            externalContext.redirect(externalContext.getRequestContextPath() + viewId);
        } catch (Exception e) {
            log.trace("Can't perform redirect to viewId: " + viewId, e);
        }
    }

}
